extension/src/lib.rs

```diff
@@ -1,324 +1,159 @@
+/// A configurable chain extension.
 #[derive(Default)]
-pub struct ApiExtension<I>(PhantomData<I>);
-
-impl<T, I> ChainExtension<T> for ApiExtension<I>
+pub struct Extension<C: Config>(PhantomData<C>);
+impl<Runtime, Config> ChainExtension<Runtime> for Extension<Config>
 where
-	T: pallet_contracts::Config
+	Runtime: pallet_contracts::Config
 		+ frame_system::Config<
 			RuntimeCall: GetDispatchInfo + Dispatchable<PostInfo = PostDispatchInfo>,
 		>,
-	T::AccountId: UncheckedFrom<T::Hash> + AsRef<[u8]>,
-	// Bound the type by the two traits which need to be implemented by the runtime.
-	I: ReadState + CallFilter<Call = <T as frame_system::Config>::RuntimeCall> + 'static,
+	Config: self::Config<Functions: Function<Config = Runtime>> + 'static,
 {
-	fn call<E: Ext<T = T>>(
-		&mut self,
-		env: Environment<E, InitState>,
-	) -> Result<RetVal, DispatchError> {
-		log::debug!(target:LOG_TARGET, " extension called ");
+	/// Call the chain extension logic.
+	///
+	/// # Parameters
+	/// - `env`: Access to the remaining arguments and the execution environment.
+	fn call<E: Ext<T = Runtime>>(&mut self, env: Environment<E, InitState>) -> Result<RetVal> {
+		log::trace!(target: Config::LOG_TARGET, "extension called");
 		let mut env = env.buf_in_buf_out();
 		// Charge weight for making a call from a contract to the runtime.
-		// `debug_message` weight is a good approximation of the additional overhead of going
-		// from contract layer to substrate layer.
+		// `debug_message` weight is a good approximation of the additional overhead of going from contract layer to substrate layer.
 		// reference: https://github.com/paritytech/ink-examples/blob/b8d2caa52cf4691e0ddd7c919e4462311deb5ad0/psp22-extension/runtime/psp22-extension-example.rs#L236
-		let contract_host_weight = ContractSchedule::<T>::get().host_fn_weights;
-		env.charge_weight(contract_host_weight.debug_message)?;
-
-		let (version, function_id, pallet_index, call_index) = extract_env(&env);
-
-		let result = match FuncId::try_from(function_id) {
-			// Read encoded parameters from buffer and calculate weight for reading `len` bytes`.
-			Ok(function_id) => {
-				// reference: https://github.com/paritytech/polkadot-sdk/blob/117a9433dac88d5ac00c058c9b39c511d47749d2/substrate/frame/contracts/src/wasm/runtime.rs#L267
-				let len = env.in_len();
-				env.charge_weight(contract_host_weight.return_per_byte.saturating_mul(len.into()))?;
-				let params = env.read(len)?;
-				match function_id {
-					FuncId::Dispatch => {
-						dispatch::<T, E, I>(&mut env, version, pallet_index, call_index, params)
-					},
-					FuncId::ReadState => {
-						read_state::<T, E, I>(&mut env, version, pallet_index, call_index, params)
-					},
-				}
-			},
-			Err(e) => Err(e),
-		};
-
-		match result {
-			Ok(_) => Ok(RetVal::Converging(0)),
-			Err(e) => Ok(RetVal::Converging(convert_to_status_code(e, version))),
+		env.charge_weight(Schedule::<Runtime>::get().host_fn_weights.debug_message)?;
```

Comment:
Different from `main`. https://github.com/r0gue-io/pop-node/blob/8072a479a3de951e560a7393e30124f91eede3a4/runtime/devnet/src/extensions.rs#L128C2-L130C60

Reply:
The read_state implementation at that reference does not seem to be consistent with the dispatch - is this intentional?

I would expect the weight charging for inputs (reading from contract memory) to be the same for both, with the read_state charging additional weight for the returned value. Am I missing something?

Reply:
The changes simply just replace the deprecated approach in `main` (which only has a basic version of the chain extension, no pop-api introduced yet) with the new one in the scope of upgrading to 1.14.0. Hence, I believe these need to be addressed in the API PR.

Reply:
Addressed in latest pushed commit.